File: server/services/commentCoverage.js
import { findDictionaryEntriesWithTextEvidence } from './deepseekKeywordTrainer.js';

function hasChinese(text) {
  return /[\p{Script=Han}]/u.test(String(text || ''));
}

function cleanComment(text) {
  return String(text || '').replace(/\s+/g, ' ').trim();
}

function stripMentionScaffolding(text) {
  return cleanComment(text)
    .replace(/回复\s*@[^:：\s]+[\s:：]*/gu, '')
    .replace(/@[^:：\s]+/gu, '')
    .trim();
}

function summarizeHit(entry) {
  return {
    term: entry.term,
    family: entry.family,
    meaning: entry.meaning,
  };
}

const EMOTE_SEMANTICS = [
  {
    pattern: /\[(?:tv_)?doge(?:[_-][^\]]+)?\]|doge|🙃|🙂|😏/iu,
    term: 'doge/反讽表情',
    family: 'attack',
    meaning: '中文平台评论中常用来标记反讽、阴阳怪气、保命式玩笑或“话里有话”的语气，不能当作普通装饰忽略。',
  },
  {
    pattern: /\[(?:藏狐|tv_斜眼笑|斜眼笑|滑稽|妙啊|阴险)\]|😅|😂|🤣|🤭/u,
    term: '嘲讽/看戏表情',
    family: 'attack',
    meaning: '表达调侃、看戏、嘲笑或讽刺态度；需要结合句子判断是玩梗还是指向具体对象的攻击。',
  },
  {
    pattern: /\[(?:吃瓜|嗑瓜子|热词系列[_-]吃瓜|doge[_-]金箍)\]|🍉/u,
    term: '吃瓜/旁观表情',
    family: 'evasion',
    meaning: '表示围观、拱火、旁观或回避直接论证，常弱化说话者责任或把严肃争论娱乐化。',
  },
  {
    pattern: /\[(?:捂脸|笑哭|喜极而泣|允悲|辣眼睛)\]|😓|🤦|🤦‍♂️|🤦‍♀️/u,
    term: '无语/尴尬表情',
    family: 'attack',
    meaning: '表达无语、尴尬、轻蔑或自嘲；在中文评论里经常承担态度和讽刺功能。',
  },
];

export function detectEmoteSemanticHits(comment) {
  const message = cleanComment(comment);
  if (!message) return [];
  return EMOTE_SEMANTICS
    .filter((item) => item.pattern.test(message))
    .map((item) => summarizeHit(item));
}

function cleanNeedle(value) {
  return String(value || '').normalize('NFKC').trim().toLowerCase();
}

function isSelfReferentialNoviceHit(entry, message) {
  if (entry?.family !== 'attack') return false;
  const term = String(entry?.term || '');
  const aliases = Array.isArray(entry?.aliases) ? entry.aliases.map(String) : [];
  if (![term, ...aliases].some((value) => value.includes('小白'))) return false;
  return /(?:^|[，,。！？!?\s])我(?:也|是|就是|也算|算)?[^，,。！？!?]{0,8}小白/u.test(message);
}

function isSuppressedLexicalHit(entry, message) {
  return isSelfReferentialNoviceHit(entry, message);
}

function exactDictionaryEntries(dictionary, message) {
  const cleanMessage = cleanNeedle(message);
  if (!cleanMessage) return [];
  const hits = [];
  for (const entry of Array.isArray(dictionary?.entries) ? dictionary.entries : []) {
    const needles = [entry.term, ...(Array.isArray(entry.aliases) ? entry.aliases : []), ...(Array.isArray(entry.examples) ? entry.examples : [])]
      .map(cleanNeedle)
      .filter((item) => item.length >= 2);
    if (needles.some((needle) => cleanMessage.includes(needle))) hits.push(entry);
  }
  return hits;
}

export function classifyCommentCoverage(dictionary, comment, options = {}) {
  const message = cleanComment(comment);
  if (!message) {
    return {
      covered: false,
      mode: 'uncovered',
      reason: 'empty comment',
      hits: [],
      comment: message,
    };
  }

  const attributableMessage = stripMentionScaffolding(message);
  const evidenceEntries = findDictionaryEntriesWithTextEvidence(dictionary, attributableMessage, {
    source: options.source || 'comment coverage check',
  }).filter((entry) => !isSuppressedLexicalHit(entry, attributableMessage));
  const lexicalEntries = evidenceEntries.length > 0
    ? evidenceEntries
    : exactDictionaryEntries(dictionary, attributableMessage)
      .filter((entry) => !isSuppressedLexicalHit(entry, attributableMessage));
  const lexicalHits = lexicalEntries.map(summarizeHit);
  const emoteHits = detectEmoteSemanticHits(message);
  const hits = [...lexicalHits, ...emoteHits];

  if (hits.length > 0) {
    return {
      covered: true,
      mode: 'keyword',
      reason: lexicalHits.length > 0 && emoteHits.length > 0
        ? 'dictionary term and emoji/emote semantic marker matched'
        : (lexicalHits.length > 0 ? 'dictionary term evidence matched' : 'emoji/emote semantic marker matched'),
      hits,
      comment: message,
    };
  }

  if (hasChinese(message)) {
    return {
      covered: true,
      mode: 'neutral',
      reason: 'no dictionary risk term matched; comment remains analyzable as neutral/no-keyword speech',
      hits: [],
      comment: message,
    };
  }

  return {
    covered: false,
    mode: 'uncovered',
    reason: 'non-Chinese or unsupported empty lexical content',
    hits: [],
    comment: message,
  };
}

export function sampleCommentCoverage(dictionary, comments = [], options = {}) {
  const sampleSize = Math.max(0, Number(options.sampleSize) || comments.length);
  const picked = comments.slice(0, sampleSize);
  const samples = picked.map((comment) => classifyCommentCoverage(dictionary, comment, options));
  const byMode = { keyword: 0, neutral: 0, uncovered: 0 };
  for (const sample of samples) {
    byMode[sample.mode] = (byMode[sample.mode] || 0) + 1;
  }
  const covered = samples.filter((sample) => sample.covered).length;
  return {
    total: samples.length,
    covered,
    uncovered: samples.length - covered,
    coverageRatio: samples.length > 0 ? covered / samples.length : 1,
    byMode,
    samples,
  };
}

File: python_backend/analysis/comment_coverage.py
from __future__ import annotations

import re
import unicodedata
from typing import Any


def _clean_text(value: Any) -> str:
    return re.sub(r"\s+", " ", str(value or "")).strip()


def _clean_needle(value: Any) -> str:
    text = unicodedata.normalize("NFKC", str(value or ""))
    return re.sub(r"[^\w\u3400-\u9fff]+", "", text, flags=re.UNICODE).lower()


def _has_chinese(value: Any) -> bool:
    return bool(re.search(r"[\u3400-\u9fff]", str(value or "")))


def _is_scrape_diagnostic(value: Any) -> bool:
    message = _clean_text(value)
    return bool(
        re.search(r"(?:^|[:\s])(?:discover|explicit Tieba thread URLs):\s+.*HTTP\s+(?:403|4\d\d|5\d\d)\s+from\s+https?://", message, re.IGNORECASE)
        or re.search(r"HTTP\s+(?:403|4\d\d|5\d\d)\s+from\s+https?://(?:tieba|c\.tieba|www\.bilibili|api\.bilibili)\.", message, re.IGNORECASE)
    )


class CommentCoverageClassifier:
    """Classify comment coverage against dictionary lexical evidence contracts."""

    def classify(self, dictionary: dict[str, Any] | None, comment: Any, options: dict[str, Any] | None = None) -> dict[str, Any]:
        message = self._comment_message(comment)
        if not message:
            return {"covered": False, "mode": "uncovered", "reason": "empty comment", "hits": [], "comment": message}
        if _is_scrape_diagnostic(message):
            return {
                "covered": True,
                "mode": "neutral",
                "reason": "scrape diagnostic line, not user speech",
                "hits": [],
                "comment": message,
            }

        hits = self._lexical_hits(dictionary or {}, message)
        if hits:
            return {
                "covered": True,
                "mode": "keyword",
                "reason": "dictionary term matched",
                "hits": hits,
                "comment": message,
            }
        if _has_chinese(message):
            return {
                "covered": True,
                "mode": "neutral",
                "reason": "no dictionary risk term matched; comment remains analyzable as neutral/no-keyword speech",
                "hits": [],
                "comment": message,
            }
        return {
            "covered": False,
            "mode": "uncovered",
            "reason": "non-Chinese or unsupported empty lexical content",
            "hits": [],
            "comment": message,
        }

    def sample(self, dictionary: dict[str, Any] | None, comments: list[Any] | None = None, options: dict[str, Any] | None = None) -> dict[str, Any]:
        options = options or {}
        comments = comments if isinstance(comments, list) else []
        sample_size = int(options.get("sampleSize") or len(comments))
        picked = comments[: max(0, sample_size)]
        samples = [self.classify(dictionary or {}, comment, options) for comment in picked]
        by_mode = {"keyword": 0, "neutral": 0, "uncovered": 0}
        for sample in samples:
            by_mode[sample["mode"]] = by_mode.get(sample["mode"], 0) + 1
        covered = len([sample for sample in samples if sample["covered"]])
        return {
            "total": len(samples),
            "covered": covered,
            "uncovered": len(samples) - covered,
            "coverageRatio": covered / len(samples) if samples else 1,
            "byMode": by_mode,
            "samples": samples,
        }

    def sample_result(self, dictionary: dict[str, Any] | None, comments: list[Any] | None = None, options: dict[str, Any] | None = None) -> dict[str, Any]:
        return {"ok": True, "summary": self.sample(dictionary, comments, options)}

    def _comment_message(self, comment: Any) -> str:
        if isinstance(comment, dict):
            return _clean_text(comment.get("message"))
        return _clean_text(comment)

    def _lexical_hits(self, dictionary: dict[str, Any], message: str) -> list[dict[str, Any]]:
        clean_message = _clean_needle(message)
        hits = []
        for entry in dictionary.get("entries") or []:
            if not isinstance(entry, dict):
                continue
            needles = [entry.get("term"), *(entry.get("aliases") or []), *(entry.get("examples") or [])]
            normalized = [_clean_needle(needle) for needle in needles if len(_clean_needle(needle)) >= 2]
            if any(needle in clean_message for needle in normalized):
                hits.append(
                    {
                        "term": _clean_text(entry.get("term")),
                        "family": _clean_text(entry.get("family") or "attack"),
                        "meaning": entry.get("meaning"),
                    }
                )
        return hits


class CommentCoverageSummary:
    """Shape comment coverage reports into the JS/Python comparator summary contract."""

    SUMMARY_KEYS = ("total", "covered", "uncovered", "coverageRatio")
    MODE_KEYS = ("keyword", "neutral", "uncovered")

    def summarize(self, summary: dict[str, Any] | None = None) -> dict[str, Any]:
        summary = summary if isinstance(summary, dict) else {}
        by_mode = summary.get("byMode") if isinstance(summary.get("byMode"), dict) else {}
        result = {key: summary.get(key) for key in self.SUMMARY_KEYS}
        result["byMode"] = {key: by_mode.get(key) for key in self.MODE_KEYS}
        return result


class CommentCoverageContractComparator:
    """Compare comment coverage summaries using the JS/Python JSON contract."""

    def __init__(self, summary: CommentCoverageSummary | None = None):
        self.summary = summary or CommentCoverageSummary()

    def compare(self, python_report: dict[str, Any] | None, js_report: dict[str, Any] | None) -> dict[str, Any]:
        python_report = python_report if isinstance(python_report, dict) else {}
        js_report = js_report if isinstance(js_report, dict) else {}
        python_summary = python_report.get("summary") if isinstance(python_report.get("summary"), dict) else {}
        js_summary = js_report.get("summary") if isinstance(js_report.get("summary"), dict) else js_report
        js_summary = js_summary if isinstance(js_summary, dict) else {}
        mismatches = self._summary_mismatches(python_summary, js_summary)
        return {
            "ok": not mismatches,
            "mismatches": mismatches,
            "python": {"summary": self.summary.summarize(python_summary)},
            "js": {"summary": self.summary.summarize(js_summary)},
        }

    def _summary_mismatches(self, python_summary: dict[str, Any], js_summary: dict[str, Any]) -> list[dict[str, Any]]:
        mismatches = [
            {"key": key, "python": python_summary.get(key), "js": js_summary.get(key)}
            for key in self.summary.SUMMARY_KEYS
            if key in js_summary and python_summary.get(key) != js_summary.get(key)
        ]
        python_modes = python_summary.get("byMode") if isinstance(python_summary.get("byMode"), dict) else {}
        js_modes = js_summary.get("byMode") if isinstance(js_summary.get("byMode"), dict) else {}
        mismatches.extend(
            {"key": f"byMode.{key}", "python": python_modes.get(key), "js": js_modes.get(key)}
            for key in self.summary.MODE_KEYS
            if key in js_modes and python_modes.get(key) != js_modes.get(key)
        )
        return mismatches
